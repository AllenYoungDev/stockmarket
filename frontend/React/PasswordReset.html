<html>
	<head>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />

		<style>
			label:after { content: ':'; }
			.top-bar {margin: 0; padding: 1% 0.5% 1% 0; background: rgb(0,0,0); color: rgb(255,255,255); text-align: left; font-family: Arial, Helvetica, sans-serif;}

			.password-reset-section {background-color: #737CA1;color: rgb(250,250,250);}
			.password-reset-section-heading {text-align: center; font-size: 250%; font-family: Arial, Helvetica, sans-serif; font-weight: normal; margin: 0;}

			.email-address {font-size: 110%;font-family: arial;text-align: center;}
			.reset-password-form-div-grid  {display:grid; grid-template-columns: 25% 25% 10% 45%; grid-gap:5px; font-size: 110%;font-family: arial;}
			.reset-password-form {margin: 1% 0% 0% 0%;}
			.reset-password-form-label-password {grid-row: 1; grid-column: 2; text-align: right;}
			.reset-password-form-input-password {grid-row: 1; grid-column: 3; text-align: left;}
			.reset-password-form-password-show-hide {grid-row: 1; grid-column: 4;margin: 0; padding: 0.35em 0 0 0.25em;text-align: left;font-size: 60%;font-family: Arial, Helvetica, sans-serif; font-weight: 50;}
			.reset-password-form-label-password-retype {grid-row: 2; grid-column: 2; text-align: right;}
			.reset-password-form-input-password-retype {grid-row: 2; grid-column: 3; text-align: left;}
			.reset-password-form-div-reset-password-button {text-align: center;}
			.reset-password-form-error-notice {text-align: center; margin: 1% 0 0 0;font-size: 120%;color: rgb(70,0,0);}
			.reset-password-form-input-reset-password-button {margin: 1% 0 1.25% 0; padding: 0.5%; font-size: 150%; font-weight: 900; background-color: #2B547E; border-radius: 0px; color: rgb(250,250,250);}
			.footer   {margin: 0; padding: 0.25% 0.5% 0.1% 0; background: rgb(0,0,0); color: rgb(255,255,255); font-family: Arial, Helvetica, sans-serif; text-align: center;}
		</style>
	</head>
  	<body style="margin: 0; background-color: #FFFFFF;color: rgb(250,250,250);">
		
		<div id="password-reset-page"></div>
		
		<script src="https://unpkg.com/react@18/umd/react.development.js" crossorigin></script>
		<script src="https://unpkg.com/react-dom@18/umd/react-dom.development.js" crossorigin></script>
		
		<script type="module" src="PasswordReset.render.js"></script>
		
		<noscript>Your browser must support JavaScript.</noscript>		

	</body>
</html>
