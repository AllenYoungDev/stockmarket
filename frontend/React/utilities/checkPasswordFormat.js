const reLetterChecker = /[a-zA-Z]/;
const reNumberChecker = /[0-9]/;
const reSpecialCharacterChecker = /[\~\`\!\@\#\$\%\^\&\*\-\_\+\=\|\\\/\?\;\:\"\'\<\>\,\.\{\}\[\]\(\)]/;

export function checkPasswordFormat(password) {
    var numberOfCharactersInPassword = password.trim().length;

    if (numberOfCharactersInPassword < 10) {return false;}

    if (password.match(reLetterChecker) === null) {return false;}

    if (password.match(reNumberChecker) === null) {return false;}

    if (password.match(reSpecialCharacterChecker) === null) {return false;}

    return true;
}