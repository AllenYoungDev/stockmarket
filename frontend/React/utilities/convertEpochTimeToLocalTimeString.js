export function convertEpochTimeToLocalTimeString(epochTime) {
    var date = new Date(epochTime);
    return date.toLocaleDateString("default") + " " + date.toLocaleTimeString("default");
}