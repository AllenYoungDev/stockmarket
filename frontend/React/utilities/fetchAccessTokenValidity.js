'use strict';

import '../settings/ServerUrls.js';

export function fetchAccessTokenValidity(setLoginState, ignore) {

    console.log('fetchAccessTokenValidity() start.')

    fetch(window.backendServerUrl + "/CheckAccessTokenValidity", {method: "GET", credentials: 'include'})
    .then((response) => response.text())
    .then((text) => {
        if (text === 'valid') {
            console.log('fetchAccessTokenValidity():  access token is valid.');
            if (ignore) {return;}
            if (document.cookie.includes('admin=true')) {
                setLoginState('admin-login');
            } else {
                setLoginState('non-admin-login');
            }
        } else {
            console.log('fetchAccessTokenValidity():  access token is invalid.');
            if (ignore) {return;}
            setLoginState('no-login');
        }
    })
    .catch((err) => {
        console.log(`fetchAccessTokenValidity():  fetch() network error.  ${err}`);
        if (ignore) {return;}
        setLoginState('fetch-error');
    });
}