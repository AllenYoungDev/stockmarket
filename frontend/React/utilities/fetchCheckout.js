import '../settings/ServerUrls.js';

export function fetchCheckout(numberOfSharesToBuy, errorNoticeDomNodeRef) {
  let checkoutResponse = 0;
  let checkoutResponseStatus = 0;

  fetch(window.backendServerUrl + "/Checkout", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },      
    body: JSON.stringify({'numberOfSharesToBuy': numberOfSharesToBuy}),
    credentials: 'include',
    })
  .then((response) => {
    checkoutResponse = response;
    response.headers.forEach((value, key) => {
      console.log(`CheckoutButton() handleOnClick() checkoutResponse.headers key==>value:  ${key} ==> ${value}`)
    });
    checkoutResponseStatus=response.status;
    return response.text();})
  .then((text) => {
    if(checkoutResponseStatus === 200) {
      //window.location.href = window.frontendServerUrl;
      //This is not the proper implementation.  The page must navigate to /Checkout URL.
    } else {
      errorNoticeDomNodeRef.current.innerHTML = `Error.  Cannot proceed to checkout.  ${text}.`
    }
  })
  .catch((err) => {
    //'fetch-error'
    errorNoticeDomNodeRef.current.innerHTML = `Unable to proceed to checkout.  Network error.`;
  });   
}