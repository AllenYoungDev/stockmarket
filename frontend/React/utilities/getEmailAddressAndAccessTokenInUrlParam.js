export function getEmailAddressAndAccessTokenInUrlParam() {
    var url_string = window.location; 
    var url = new URL(url_string);
    var emailAddress = url.searchParams.get("emailAddress");
    var accessToken = url.searchParams.get("accessToken");
    console.log(`PasswordResetControls() accessToken:  ${accessToken}.`);
    return [emailAddress, accessToken];
}