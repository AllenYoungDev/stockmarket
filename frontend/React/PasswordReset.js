'use strict';

//import React, { StrictMode } from "react";
//import { createRoot } from 'react-dom/client';

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

import './settings/ServerUrls.js';
import { fetchResetPassword } from './utilities/fetchResetPassword.js';
import { getEmailAddressAndAccessTokenInUrlParam } from './utilities/getEmailAddressAndAccessTokenInUrlParam.js';
import { checkPasswordFormat } from './utilities/checkPasswordFormat.js';
import { footer } from './footer.js';

/* *************************************************************************************************************
PasswordResetPage React component
************************************************************************************************************** */
export function PasswordResetPage() {
  return React.createElement(
    React.Fragment,
    null,
    React.createElement(
      'div',
      { className: 'top-bar' },
      React.createElement(
        'a',
        { href: window.frontendServerUrl, style: { textDecoration: 'none', color: 'rgb(255,255,255)' } },
        'Allen Young\'s Stockmarket Demo'
      )
    ),
    React.createElement(
      'section',
      { className: 'password-reset-section' },
      React.createElement(
        'h1',
        { className: 'password-reset-section-heading' },
        'Password reset'
      ),
      React.createElement(PasswordResetControls, null)
    ),
    footer()
  );
}

/* *************************************************************************************************************
PasswordResetControls React component
************************************************************************************************************** */

export function PasswordResetControls() {
  var passwordInputRef = React.useRef(null);
  var passwordRetypeInputRef = React.useRef(null);
  var errorNoticeDomNodeRef = React.useRef(null);

  var _getEmailAddressAndAc = getEmailAddressAndAccessTokenInUrlParam(),
      _getEmailAddressAndAc2 = _slicedToArray(_getEmailAddressAndAc, 2),
      emailAddress = _getEmailAddressAndAc2[0],
      accessToken = _getEmailAddressAndAc2[1];

  function handleShowHideClick(event) {
    event.preventDefault();
    event.stopPropagation();

    if (event.target.innerHTML === 'show') {
      event.target.innerHTML = 'hide';
      passwordInputRef.current.type = "text";
      passwordRetypeInputRef.current.type = "text";
    } else {
      event.target.innerHTML = 'show';
      passwordInputRef.current.type = "password";
      passwordRetypeInputRef.current.type = "password";
    }
  }

  function handleOnClick(event) {
    event.stopPropagation();

    errorNoticeDomNodeRef.current.innerHTML = '';

    passwordInputRef.current.style.backgroundColor = '#FFFFFF';
    passwordRetypeInputRef.current.style.backgroundColor = '#FFFFFF';

    /* ************************************************************************************************
    Input validation
    ************************************************************************************************ */

    var password = passwordInputRef.current.value.trim();
    if (password === '') {
      errorNoticeDomNodeRef.current.innerHTML = "Unable to reset password.  You need to enter password.  Please correct and try again.";
      passwordInputRef.current.style.backgroundColor = '#FAA0A0';
      return;
    }

    if (!checkPasswordFormat(password)) {
      errorNoticeDomNodeRef.current.innerHTML = "Unable to reset password.  Password does not meet the requirement.  Please correct and try again.";
      passwordInputRef.current.style.backgroundColor = '#FAA0A0';
      return;
    }

    var passwordRetype = passwordRetypeInputRef.current.value.trim();
    if (passwordRetype === '') {
      errorNoticeDomNodeRef.current.innerHTML = "Unable to reset password.  You need to enter password (retype).  Please correct and try again.";
      passwordRetypeInputRef.current.style.backgroundColor = '#FAA0A0';
      return;
    }

    if (passwordRetype !== password) {
      errorNoticeDomNodeRef.current.innerHTML = "Password and password retype do not match.  Please correct and try again.";
      passwordInputRef.current.style.backgroundColor = '#FAA0A0';
      passwordRetypeInputRef.current.style.backgroundColor = '#FAA0A0';
      return;
    }

    /* ************************************************************************************************
    Password reset submission to backend
    ************************************************************************************************ */
    fetchResetPassword(accessToken, password, errorNoticeDomNodeRef);
  }

  return React.createElement(
    React.Fragment,
    null,
    React.createElement(
      'p',
      { className: 'email-address' },
      'Use the form below to change your login password.'
    ),
    React.createElement(
      'p',
      { className: 'email-address' },
      'Email address:  ',
      emailAddress
    ),
    React.createElement(
      'form',
      { className: 'reset-password-form' },
      React.createElement(
        'div',
        { className: 'reset-password-form-div-grid' },
        React.createElement(
          'label',
          { className: 'reset-password-form-label-password', htmlFor: 'password' },
          'Password'
        ),
        React.createElement('input', { className: 'reset-password-form-input-password', type: 'password', id: 'password', name: 'password', ref: passwordInputRef }),
        React.createElement('br', null),
        React.createElement(
          'p',
          { className: 'reset-password-form-password-show-hide' },
          React.createElement(
            'a',
            { href: '', role: 'password-show-hide', style: { textDecoration: 'none', color: 'rgb(255,255,255)' }, onClick: handleShowHideClick },
            'show'
          )
        ),
        React.createElement(
          'label',
          { className: 'reset-password-form-label-password-retype', htmlFor: 'passwordretype' },
          'Password (Retype)'
        ),
        React.createElement('input', { className: 'reset-password-form-input-password-retype', type: 'password', id: 'passwordretype', name: 'passwordretype', size: '20', ref: passwordRetypeInputRef })
      ),
      React.createElement(
        'div',
        { className: 'reset-password-form-div-reset-password-button' },
        React.createElement('p', { className: 'reset-password-form-error-notice', role: 'error-notice', ref: errorNoticeDomNodeRef }),
        React.createElement('input', { className: 'reset-password-form-input-reset-password-button', type: 'button', value: 'Reset password', onClick: handleOnClick })
      )
    )
  );
}