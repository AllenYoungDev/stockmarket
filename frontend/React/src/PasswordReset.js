'use strict';

//import React, { StrictMode } from "react";
//import { createRoot } from 'react-dom/client';

import './utilities/ServerUrl.js';
import { fetchResetPassword } from './utilities/fetchResetPassword.js';
import { getEmailAddressAndAccessTokenInUrlParam } from './utilities/getEmailAddressAndAccessTokenInUrlParam.js';
import { checkPasswordFormat } from './utilities/checkPasswordFormat.js';
import {footer} from './footer.js';

/* *************************************************************************************************************
PasswordResetPage React component
************************************************************************************************************** */
export function PasswordResetPage() {
  return (
    <React.Fragment>
    <div className="top-bar">
    <a href={window.frontendServerUrl} style={{textDecoration: 'none', color: 'rgb(255,255,255)'}}>Allen Young's Stockmarket Demo</a>
    </div>
    
		<section className="password-reset-section">
		<h1 className="password-reset-section-heading">Password reset</h1>
    <PasswordResetControls />
    </section>

    {footer()}
    </React.Fragment>
  );
}

/* *************************************************************************************************************
PasswordResetControls React component
************************************************************************************************************** */

export function PasswordResetControls() {
  const passwordInputRef = React.useRef(null);
  const passwordRetypeInputRef = React.useRef(null);  
  const errorNoticeDomNodeRef = React.useRef(null);

  const [emailAddress, accessToken] = getEmailAddressAndAccessTokenInUrlParam();
  
  function handleShowHideClick(event) {
    event.preventDefault();
    event.stopPropagation();

    if (event.target.innerHTML === 'show') {
      event.target.innerHTML = 'hide';
      passwordInputRef.current.type = "text";
      passwordRetypeInputRef.current.type = "text";
    } else {
      event.target.innerHTML = 'show';
      passwordInputRef.current.type = "password";
      passwordRetypeInputRef.current.type = "password";
    }
  }

  function handleOnClick(event) {
    event.stopPropagation();

    errorNoticeDomNodeRef.current.innerHTML = '';

    passwordInputRef.current.style.backgroundColor = '#FFFFFF';
    passwordRetypeInputRef.current.style.backgroundColor = '#FFFFFF';   

    /* ************************************************************************************************
    Input validation
    ************************************************************************************************ */

    const password = passwordInputRef.current.value.trim();
    if(password === '') {
      errorNoticeDomNodeRef.current.innerHTML = "Unable to reset password.  You need to enter password.  Please correct and try again.";
      passwordInputRef.current.style.backgroundColor = '#FAA0A0';
      return
    }

    if(!checkPasswordFormat(password)) {
      errorNoticeDomNodeRef.current.innerHTML = "Unable to reset password.  Password does not meet the requirement.  Please correct and try again.";
      passwordInputRef.current.style.backgroundColor = '#FAA0A0';
      return
    }

    const passwordRetype = passwordRetypeInputRef.current.value.trim();
    if(passwordRetype === '') {
      errorNoticeDomNodeRef.current.innerHTML = "Unable to reset password.  You need to enter password (retype).  Please correct and try again.";
      passwordRetypeInputRef.current.style.backgroundColor = '#FAA0A0'; 
      return
    }

    if(passwordRetype !== password) {
      errorNoticeDomNodeRef.current.innerHTML = "Password and password retype do not match.  Please correct and try again.";
      passwordInputRef.current.style.backgroundColor = '#FAA0A0';
      passwordRetypeInputRef.current.style.backgroundColor = '#FAA0A0';      
      return
    }

    /* ************************************************************************************************
    Password reset submission to backend
    ************************************************************************************************ */
    fetchResetPassword(accessToken, password, errorNoticeDomNodeRef);
  }

  return (
    <React.Fragment>
      <p className="email-address">Use the form below to change your login password.</p>

    <p className="email-address">Email address:  {emailAddress}</p>

    <form className="reset-password-form">
      <div className="reset-password-form-div-grid">
      <label className="reset-password-form-label-password" htmlFor="password">Password</label>
      <input className="reset-password-form-input-password" type="password" id="password" name="password" ref={passwordInputRef} /><br />
      <p className="reset-password-form-password-show-hide"><a href="" role="password-show-hide" style={{textDecoration: 'none', color: 'rgb(255,255,255)'}} onClick={handleShowHideClick}>show</a></p>
      <label className="reset-password-form-label-password-retype" htmlFor="passwordretype">Password (Retype)</label>
      <input className="reset-password-form-input-password-retype" type="password" id="passwordretype" name="passwordretype" size="20" ref={passwordRetypeInputRef} />   
      </div>

      <div className="reset-password-form-div-reset-password-button">
      <p className="reset-password-form-error-notice" role="error-notice" ref={errorNoticeDomNodeRef}></p>
      <input className="reset-password-form-input-reset-password-button" type="button" value="Reset password" onClick={handleOnClick} />
      </div>
    </form>
    </React.Fragment>
  );
}
