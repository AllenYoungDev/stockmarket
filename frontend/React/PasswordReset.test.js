/**
 * @jest-environment jsdom
 */

//import React from "react";
//import ReactDOM from 'react-dom/client';

import renderer from 'react-test-renderer';
import {PasswordResetPage} from './PasswordReset.js';

jest.mock('./utilities/getEmailAddressAndAccessTokenInUrlParam');

import {getEmailAddressAndAccessTokenInUrlParam} from './utilities/getEmailAddressAndAccessTokenInUrlParam';

(async () => {
  if (typeof React === 'undefined') {
    globalThis.React = await import("react");
  }
  if (typeof ReactDOM === 'undefined') {
    globalThis.ReactDOM = await import("react-dom/client");
  }
})();


it('React app Jest snapshot testing for Allen Young\'s Stockmarket Demo password reset page', async () => {

  getEmailAddressAndAccessTokenInUrlParam.mockImplementation(() => {
    console.log('getEmailAddressAndAccessTokenInUrlParam() mock called!');
    return ['test@allenyoung.dev', '1111-1111-1111-1111']
  });


  let testRenderer;
  let tree;

  testRenderer = renderer.create(
    <PasswordResetPage />,
  );

  tree = testRenderer.toJSON();
  expect(tree).toMatchSnapshot();
});