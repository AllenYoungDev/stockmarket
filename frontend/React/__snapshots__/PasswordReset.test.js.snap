// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React app Jest snapshot testing for Allen Young's Stockmarket Demo password reset page 1`] = `
[
  <div
    className="top-bar"
  >
    <a
      href="http://localhost"
      style={
        {
          "color": "rgb(255,255,255)",
          "textDecoration": "none",
        }
      }
    >
      Allen Young's Stockmarket Demo
    </a>
  </div>,
  <section
    className="password-reset-section"
  >
    <h1
      className="password-reset-section-heading"
    >
      Password reset
    </h1>
    <p
      className="email-address"
    >
      Use the form below to change your login password.
    </p>
    <p
      className="email-address"
    >
      Email address:  
      test@allenyoung.dev
    </p>
    <form
      className="reset-password-form"
    >
      <div
        className="reset-password-form-div-grid"
      >
        <label
          className="reset-password-form-label-password"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="reset-password-form-input-password"
          id="password"
          name="password"
          type="password"
        />
        <br />
        <p
          className="reset-password-form-password-show-hide"
        >
          <a
            href=""
            onClick={[Function]}
            style={
              {
                "color": "rgb(255,255,255)",
                "textDecoration": "none",
              }
            }
          >
            show
          </a>
        </p>
        <label
          className="reset-password-form-label-password-retype"
          htmlFor="passwordretype"
        >
          Password (Retype)
        </label>
        <input
          className="reset-password-form-input-password-retype"
          id="passwordretype"
          name="passwordretype"
          size="20"
          type="password"
        />
      </div>
      <div
        className="reset-password-form-div-reset-password-button"
      >
        <p
          className="reset-password-form-error-notice"
        />
        <input
          className="reset-password-form-input-reset-password-button"
          onClick={[Function]}
          type="button"
          value="Reset password"
        />
      </div>
    </form>
  </section>,
  <footer
    className="footer"
  >
    <p>
      Allen Young's Stockmarket Demo by Allen Young.
    </p>
    <p>
      For demonstrating Allen Young's full-stack web app development skills to potential employers.
    </p>
    <p>
      For more of Allen Young's software development skills demonstration, visit 
      <a
        href="https://AllenYoung.dev"
        style={
          {
            "color": "white",
          }
        }
        target="_blank"
      >
        AllenYoung.dev
      </a>
       and 
      <a
        href="https://GitHub.com/AllenYoungDev"
        style={
          {
            "color": "white",
          }
        }
        target="_blank"
      >
        GitHub.com/AllenYoungDev
      </a>
      .
    </p>
    <p>
      Â© 2023 Allen Young.  All rights reserved.
    </p>
  </footer>,
]
`;
